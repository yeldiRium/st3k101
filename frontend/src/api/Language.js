import {map} from "ramda";

import {fetchApi} from "./Util/Request";
import {extractJson} from "./Util/Response";
import {parseLanguage} from "./Util/Parse";

/**
 * Fetches all available Languages on the server.
 *
 * @returns {Future}
 * @resolve {Array<Language>}
 * @reject {TypeError|ApiError}
 * @cancel
 */
function fetchLanguages() {
    return fetchApi(
        "/api/language", {}
    )
        .chain(extractJson)
        .map(map(parseLanguage))
}

export {
    fetchLanguages
};
