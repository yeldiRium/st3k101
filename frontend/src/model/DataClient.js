import Party from "./Party";

class DataClient extends Party {
    constructor(href,
                email,
                language) {
        super(href);
        this._email = email;
        this._language = language;
    }

    // TODO: refactor email out into own model and/or validate

    /**
     * @param {string} email
     */
    set email(email) {
        this._email = email;
    }

    /**
     * @returns {string}
     */
    get email() {
        return this._email;
    }

    // TODO: refactor language out into own model and/or validate

    /**
     * @param {string} language
     */
    set language(language) {
        this._language = language;
    }

    /**
     * @returns {string}
     */
    get language() {
        return this._language;
    }
}

export default DataClient;

export {
    DataClient
};