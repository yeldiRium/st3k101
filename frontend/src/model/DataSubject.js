class DataSubject {
  // does not extend Party, as this is not a real resource
  /**
   * A user who submits data to the survey platform.
   *
   * @param {string} id
   * @param {string} email
   * @param {string} ltiUserId
   * @param {string} moodleUsername
   * @param {string} source The source LMS.
   */
  constructor(id, email, ltiUserId, moodleUsername, source) {
    this._id = id;
    this._email = email;
    this._ltiUserId = ltiUserId;
    this._moodleUsername = moodleUsername;
    this._source = source;
  }

  clone() {
    return new DataSubject(
      this._id,
      this._email,
      this._ltiUserId,
      this._moodleUsername,
      this._source
    );
  }

  get id() {
    return this._id;
  }

  get source() {
    return this._source;
  }

  get moodleUsername() {
    return this._moodleUsername;
  }

  get ltiUserId() {
    return this._ltiUserId;
  }

  get email() {
    return this._email;
  }
}

export default DataSubject;

export { DataSubject };
