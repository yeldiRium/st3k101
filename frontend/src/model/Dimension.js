import Future from "fluture";

import Party from "./Party";
import Question, {populateOwnedIncomingReferences} from "./Question";
import SurveyBase from "./SurveyBase";

class Dimension extends SurveyBase {
    /**
     * @param {string} href See Resource.
     * @param {Party}  owner See OwnedResource.
     * @param {LanguageData} languageData See SurveyBase.
     * @param {string} name The Dimension's name.
     * @param {Array.<Question>} questions An Array of all connected Questions.
     * @param {boolean} randomizeQuestions Whether the Questions should be dis-
     *  played in a random order to DataSubjects.
     */
    constructor(href,
                owner,
                languageData,
                name,
                questions,
                randomizeQuestions) {
        super(href, owner, languageData);
        this.name = name;
        this.questions = questions;
        this.randomizeQuestions = randomizeQuestions;
    }

    /**
     * True, if this Question is a ShadowQuestion.
     *
     * @returns {boolean}
     */
    get isShadow() {
        throw new Error("Please override this.");
    };

    /**
     * True, if this Question is a ConcreteQuestion.
     *
     * @returns {boolean}
     */
    get isConcrete() {
        throw new Error("Please override this.");
    }
}

class ConcreteDimension extends Dimension {
    /**
     *
     * @param {string} href See Resource.
     * @param {Party}  owner See OwnedResource.
     * @param {LanguageData} languageData See SurveyBase.
     * @param {string} name See Dimension.
     * @param {Array.<Question>} questions See Dimension.
     * @param {boolean} randomizeQuestions See Dimension.
     * @param {number}  incomingReferenceCount Number of references to this
     *  Dimension.
     *  This counts references not owned by the current user and can thus be
     *  bigger than the number of ownedIncomingReferences.
     * @param {Array.<Resource|ShadowDimension>} ownedIncomingReferences List of
     *  all references (in form of hrefs or ShadowDimension instances) to this
     *  Dimension, which the current user owns.
     */
    constructor(href,
                owner,
                languageData,
                name,
                questions,
                randomizeQuestions,
                incomingReferenceCount,
                ownedIncomingReferences) {
        super(href, owner, languageData, name, questions, randomizeQuestions);

        this.incomingReferenceCount = incomingReferenceCount;
        this.ownedIncomingReferences = ownedIncomingReferences;
    }

    get isShadow() {
        return false;
    }

    get isConcrete() {
        return true;
    }

    /**
     * Number of reference to this Dimension, which are owned by the current
     * user.
     *
     * @returns {number}
     */
    ownedIncomingReferenceCount() {
        return this.ownedIncomingReferences.length;
    }
}

class ShadowDimension extends Dimension {
    /**
     *
     * @param {string} href See Resource.
     * @param {Party}  owner See OwnedResource.
     * @param {LanguageData} languageData See SurveyBase.
     * @param {string} name See Dimension.
     * @param {Array.<Question>} questions See Dimension.
     * @param {boolean} randomizeQuestions See Dimension.
     * @param {Resource|ConcreteDimension} referenceTo Href or instance of the
     *  referenced Dimension.
     */
    constructor(href,
                owner,
                languageData,
                name,
                questions,
                randomizeQuestions,
                referenceTo) {
        super(href, owner, languageData, name, questions, randomizeQuestions);
        this.referenceTo = referenceTo;
    }

    get isShadow() {
        return true;
    }

    get isConcrete() {
        return false;
    }
}

/**
 * Takes in a Dimension and populates its incoming references field by replacing
 * all resolvable References with their corresponding ShadowDimension instances.
 *
 * Accesses the API to load the Dimensions.
 *
 * TODO: implement. currently this replaces all Resource-type references with
 *       empty objects.
 *
 * @param {ConcreteDimension} concreteDimension
 * @return {Future}
 * @resolve with nothing, since the change is made in place.
 * @reject with an API error message.
 * @cancel TODO: can this actually be cancelled?
 */
function populateOwnedIncomingReferences(concreteDimension) {
    const resolvedShadowDimensionFutures = [];

    // Use basic for loop to easily replace values.
    for (let i = 0; i < concreteDimension.ownedIncomingReferences.length; i++) {
        let reference = concreteDimension.ownedIncomingReferences[i];

        if (instanceOf(reference, Resource)) {
            // TODO: fetch ShadowDimension and replace
            const shadowDimensionFuture = Future.of({}) // <- replace with actual API call
                .map(shadowDimension => {
                    concreteDimension.ownedIncomingReferences[i] =
                        shadowDimension;
                    return true;
                });

            resolvedShadowDimensionFutures.push(shadowDimensionFuture);
        }
    }
    return Future.parallel(Infinity, resolvedShadowDimensionFutures);
}

/**
 * If the referenceTo field contains a Resource, it is resolved to a
 * ConcreteDimension instance. Otherwise it is left as is.
 *
 * Accesses the API to load the Dimension.
 *
 * If it is invalid or does not resolve to anything, an error is thrown.
 *
 * TODO: implement this. currently it replaces the reference with an empty
 *       object.
 *
 * @param {ShadowDimension} shadowDimension
 * @return {Future}
 * @resolve with nothing, since the change is made in place.
 * @reject with an API error message.
 * @cancel TODO: can this actually be cancelled?
 */
function populateReferenceTo(shadowDimension) {
    if (instanceOf(shadowDimension.referenceTo, Resource)) {
        // TODO: fetch ConcreteDimension and replace
        return Future.of({}) // <- replace with actual API call
            .map(concreteDimension => {
                shadowDimension.referenceTo = concreteDimension;
            });
    }
    return Future.of(true);
}

/**
 * Based on the type of the given Dimension this either populates its incoming
 * references or its referenceTo field.
 *
 * @param {Dimension} dimension
 * @return {Future}
 * @resolve with nothing, since the change is made in place.
 * @reject with an API error message.
 * @cancel TODO: can this actually be cancelled?
 */
function populateDimension(dimension) {
    if (instanceOf(dimension, ShadowDimension)) {
        return populateReferenceTo(dimension);
    }
    if (instanceOf(dimension, ConcreteDimension)) {
        return populateOwnedIncomingReferences(dimension);
    }
}

export default Dimension;

export {
    Dimension,
    ConcreteDimension,
    ShadowDimension,
    populateOwnedIncomingReferences,
    populateReferenceTo,
    populateDimension
}