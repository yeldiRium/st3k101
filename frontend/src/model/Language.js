class Language {
    constructor(shortName, longName) {
        this.shortName = shortName;
        this.longName = longName;
    }

    /**
     * @param {Language} language
     * @return {Boolean}
     */
    equals(language) {
        return this.shortName == language.shortName
            && this.longName == language.longName;
    }
}

class LanguageData {
    /**
     * Each in short version.
     *
     * @param {Language} currentLanguage
     * @param {Language} originalLanguage
     * @param {Array.<Language>} availableLanguages
     */
    constructor(currentLanguage, originalLanguage, availableLanguages) {
        this.currentLanguage = currentLanguage;
        this.originalLanguage = originalLanguage;
        this.availableLanguages = availableLanguages;
    }

    clone() {
        return new LanguageData(
            this.currentLanguage,
            this.originalLanguage,
            [...this.availableLanguages]
        );
    }
}

export default Language;

export {
    Language,
    LanguageData
}
