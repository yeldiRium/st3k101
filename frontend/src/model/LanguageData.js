class LanguageData {
    /**
     * Each in short version.
     *
     * @param {string} currentLanguage
     * @param {string} originalLanguage
     * @param {Array.<string>} availableLanguages
     */
    constructor(currentLanguage, originalLanguage, availableLanguages) {
        this.currentLanguage = currentLanguage;
        this.originalLanguage = originalLanguage;
        this.availableLanguages = availableLanguages;
    }
}

export default LanguageData;