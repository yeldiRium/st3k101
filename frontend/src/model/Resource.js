/**
 * An object that has a representation in the API and can be constructed by
 * fetching the given href.
 */
class Resource {
    /**
     *
     * @param {string} href
     */
    constructor(href, id) {
        this._href = href;
        this._id = id;
    }

    /**
     *
     * @returns {string}
     */
    get href() {
        return this._href;
    }

    /**
     *
     * @returns {String}
     */
    get id() {
        return this._id;
    }

    /**
     * @param {Resource} otherResource
     * @return {boolean} If the Resources are equal.
     */
    identifiesWith(otherResource) {
        return this.href === otherResource.href;
    }
}

/**
 * @param {Resource} aResource
 * @param {Resource} bResource
 * @returns {boolean} True, if a identifies with b.
 */
function identifiesWith(aResource, bResource) {
    return aResource.identifiesWith(bResource);
}

export default Resource;

export {
    Resource,
    identifiesWith
};
