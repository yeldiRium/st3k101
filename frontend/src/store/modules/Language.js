import {__, allPass, has, map} from "ramda";
import Future from "fluture";

import Language from "../../api/Model/Language";

const store = {
    namespaced: true,
    state: {
        loading: {
            state: "loading",
            error: null
        },
        currentLanguage: {
            short: "de",
            long: "Deutsch"
        },
        languageOptions: []
    },
    getters: {
        loading: state => {
            return {
                loadingState: state.loading.state,
                error: state.loading.error
            };
        },
    },
    actions: {
        /**
         * Fetch the available Languages from the API and parse them to be
         * usable by the dropdown menu.
         */
        fetchLanguages({commit}) {
            return Language.all()
                .mapRej(error => {
                    commit(
                        "setLoadingState", {
                            loadingState: "error",
                            error: error
                        }
                    );
                    return error;
                })
                .map(
                    data => {
                        data = map(
                            languageTuple => ({
                                short: languageTuple[0],
                                long: languageTuple[1]
                            }),
                            data
                        );
                        commit(
                            "setLoadingState", {
                                loadingState: "done"
                            }
                        );
                        commit("setLanguageOptions", data);

                        return data;
                    }
                );
        }
    },
    mutations: {
        setLoadingState(state, {loadingState, error}) {
            state.loading.state = loadingState;
            state.loading.error = error;
        },
        setCurrentLanguage(state, language) {
            state.currentLanguage = language;
        },
        setLanguageOptions(state, languageOptions) {
            state.languageOptions = languageOptions;
        }
    }
};

export default store;

/**
 * Initializes this store's content.
 *
 * Since actions encase their return value in a resolved Promise (if the value
 * isn't a Promise itself), the Promise's resolving value has to be untangled.
 *
 * @param rootStore The instantiated store.
 * @param namespace The namespace in which this store resides.
 * @return a Future
 * @resolves with nothing.
 * @rejects with nothing.
 * @cancel doesn't exist.
 */
const initialize = function (rootStore, namespace) {
    return Future((reject, resolve) => {
        Future.tryP(
            () => rootStore.dispatch(`${namespace}/fetchLanguages`)
        )
        // The Promise return from the fetchLanguages action will always re-
        // solve to a Future.
            .value(resolve);
    });
};

export {
    store,
    initialize
}