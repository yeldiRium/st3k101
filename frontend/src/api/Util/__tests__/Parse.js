import * as R from "ramda";

import {
  parseLanguage,
  parseLanguageData,
  parseResource,
  parseRoles
} from "../Parse";
import { Language, LanguageData } from "../../../model/Language";
import { Resource } from "../../../model/Resource";

describe("parseLanguage", () => {
  const testParams = {
    item_id: "testItemId",
    value: "testValue"
  };

  test("raises error for missing properties", () => {
    R.forEach(
      key =>
        expect(() => {
          parseLanguage(R.dissoc(key, testParams));
        }).toThrowErrorMatchingSnapshot(),
      R.keys(testParams)
    );
  });

  test("build Language object from given object", () => {
    expect(parseLanguage(testParams)).toEqual(
      new Language(testParams.item_id, testParams.value)
    );
  });
});

describe("parseLanguageData", () => {
  const testParams = {
    current_language: {
      item_id: "testItemId",
      value: "testValue"
    },
    original_language: {
      item_id: "testItemId",
      value: "testValue"
    },
    available_languages: [
      {
        item_id: "testItemId",
        value: "testValue"
      },
      {
        item_id: "testItemId",
        value: "testValue"
      }
    ]
  };

  test("raises error for missing properties", () => {
    R.forEach(
      key =>
        expect(() => {
          parseLanguageData(R.dissoc(key, testParams));
        }).toThrowErrorMatchingSnapshot(),
      R.keys(testParams)
    );
  });

  test("build LanguageData object from given object", () => {
    expect(parseLanguageData(testParams)).toEqual(
      new LanguageData(
        new Language(
          testParams.current_language.item_id,
          testParams.current_language.value
        ),
        new Language(
          testParams.original_language.item_id,
          testParams.original_language.value
        ),
        [
          new Language(
            testParams.available_languages[0].item_id,
            testParams.available_languages[0].value
          ),
          new Language(
            testParams.available_languages[1].item_id,
            testParams.available_languages[1].value
          )
        ]
      )
    );
  });
});

describe("parseResource", () => {
  const testParams = {
    href: "testHref",
    id: "testId"
  };

  test("raises error for missing properties", () => {
    R.forEach(
      key =>
        expect(() => {
          parseResource(R.dissoc(key, testParams));
        }).toThrowErrorMatchingSnapshot(),
      R.keys(testParams)
    );
  });

  test("build Language object from given object", () => {
    expect(parseResource(testParams)).toEqual(
      new Resource(testParams.href, testParams.id)
    );
  });
});

describe("parseRoles", () => {
  const testParams = [
    {
      value: "testValue"
    },
    {
      value: "testValue"
    },
    {
      value: "testValue"
    }
  ];

  test("raises error for missing properties", () => {
    expect(() => {
      parseRoles([{}, {}]);
    }).toThrowErrorMatchingSnapshot();
  });

  test("build Language object from given object", () => {
    expect(parseRoles(testParams)).toEqual([
      "testValue",
      "testValue",
      "testValue"
    ]);
  });
});

describe("parseDataClient", () => {});

describe("parseSmallDataClient", () => {});

describe("parseDataSubject", () => {});

describe("parseChallenges", () => {});

describe("parseEMailWhitelistChallenge", () => {});

describe("parseEmailBlacklistChallenge", () => {});

describe("parsePasswordChallenge", () => {});

describe("parseQuestionnaire", () => {});

describe("parseTemplateQuestionnaire", () => {});

describe("parseShadowQuestionnaire", () => {});

describe("parseConcreteQuestionnaire", () => {});

describe("parseDimension", () => {});

describe("parseTemplateDimension", () => {});

describe("parseShadowDimension", () => {});

describe("parseConcreteDimension", () => {});

describe("parseQuestion", () => {});

describe("parseTemplateQuestion", () => {});

describe("parseShadowQuestion", () => {});

describe("parseConcreteQuestion", () => {});

describe("parsePropertyUpdatedTrackerEntry", () => {});

describe("parseTranslatedPropertyUpdatedTrackerEntry", () => {});

describe("parseItemAddedTrackeEntry", () => {});

describe("parseItemRemovedTrackerEntry", () => {});

describe("parseQuestionnaireRemovedTrackerEntry", () => {});

describe("parseTrackerEntry", () => {});

describe("parseSubmissionQuestion", () => {});

describe("parseSubmissionDimension", () => {});

describe("parseSubmissionQuestionnaire", () => {});

describe("parseQuestionStatistic", () => {});
