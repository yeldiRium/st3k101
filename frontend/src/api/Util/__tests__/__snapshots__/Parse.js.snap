// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseDataClient raises error for missing properties 1`] = `"Field \`email\` is required."`;

exports[`parseDataClient raises error for missing properties 2`] = `"Field \`id\` is required."`;

exports[`parseDataClient raises error for missing properties 3`] = `"Field \`href\` is required."`;

exports[`parseDataClient raises error for missing properties 4`] = `"Field \`language\` is required."`;

exports[`parseDataClient raises error for missing properties 5`] = `"Field \`roles\` is required."`;

exports[`parseLanguage raises error for missing properties 1`] = `"Field \`item_id\` is required."`;

exports[`parseLanguage raises error for missing properties 2`] = `"Field \`value\` is required."`;

exports[`parseLanguageData raises error for missing properties 1`] = `"Field \`current_language\` is required."`;

exports[`parseLanguageData raises error for missing properties 2`] = `"Field \`original_language\` is required."`;

exports[`parseLanguageData raises error for missing properties 3`] = `"Field \`available_languages\` is required."`;

exports[`parseResource raises error for missing properties 1`] = `"Field \`href\` is required."`;

exports[`parseResource raises error for missing properties 2`] = `"Field \`id\` is required."`;

exports[`parseRoles raises error for missing properties 1`] = `"Field \`value\` is required on role."`;
